File: study/ts/script/ui/component/CommonStoryAvatar.ts
import { ConfigNameConst } from "../../const/ConfigNameConst";
import { G_ConfigLoader } from "../../init";
import { Path } from "../../utils/Path";
import { TypeConvertHelper } from "../../utils/TypeConvertHelper";
import UIHelper from "../../utils/UIHelper";
import { SpineNode } from "../node/SpineNode";
import { StorySpineNode } from "../node/StorySpineNode";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonStoryAvatar extends cc.Component {

    @property({ type: cc.Node, visible: true })
    _imageAvatar: cc.Node = null;

    @property({ type: cc.Node, visible: true })
    _nodeAvatar: cc.Node = null;

    private _spine: SpineNode | StorySpineNode;

    public updateUI(heroId, limitLevel?, limitRedLevel?) {
        this._imageAvatar.active = false;
        var param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_HERO, heroId, null, null, limitLevel, limitRedLevel);
        var resData = param.res_cfg;
        if (resData.story_res_spine != 0) {
            this._createHeroSpine(resData.story_res_spine);
            this._imageAvatar.active = (false);
        } else {
            this._createHeroImage(resData.story_res);
            if (this._spine) {
                this._spine.node.active = (false);
            }
        }
    }

    public onEnable() {
        if (this._spine) {
            this._spine.setAnimation('idle', true);
        }
    }

    private _createHeroSpine(spineId) {
        this._createSpine(Path.getStorySpine(spineId), true);
    }
    private _createHistorySpine(spineId) {
        this._createSpine(Path.getSpine(spineId), false);
    }

    public playIdleAni(): void {
        this._spine.setAnimation('idle', true);
    }

    private _createSpine(spinePath: string, isStorySpine: boolean) {
        if (this._spine == null) {
            if (isStorySpine) {
                this._spine = StorySpineNode.create(1, cc.size(1024, 1024))
            }
            else {
                this._spine = SpineNode.create(1, cc.size(1024, 1024));
            }
            this._nodeAvatar.addChild(this._spine.node);
        }
        this._spine.setAsset(spinePath);
        this._spine.setAnimation('idle', true);
        this._spine.node.active = (true);
    }

    private _createHeroImage(imageId) {
        var imgPath = Path.getChatRoleRes(imageId);
        let sprite: cc.Sprite = this._imageAvatar.getComponent(cc.Sprite);
        sprite.sizeMode = cc.Sprite.SizeMode.RAW;
        UIHelper.loadTexture(sprite, imgPath);
        this._imageAvatar.active = (true);
    }

    public updateUIByResId(resId) {
        var resData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_RES).get(resId);
        if (resData.story_res_spine != 0) {
            this._createHistorySpine(resData.story_res_spine);
            this._imageAvatar.active = (false);
        } else {
            this._createHeroImage(resData.story_res);
            if (this._spine) {
                this._spine.node.active = (false);
            }
        }
    }
}

File: study/lua/src/app/ui/component/CommonStoryAvatar.lua
--通用武将立绘spine
local CommonStoryAvatar = class("CommonStoryAvatar")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local HeroRes = require("app.config.hero_res")

local EXPORTED_METHODS = {
    "updateUI",
    "updateUIByResId"
}

function CommonStoryAvatar:ctor()
    self._target = nil
    self._spine = nil
end


function CommonStoryAvatar:_init()
    self._imageAvatar = ccui.Helper:seekNodeByName(self._target, "ImageAvatar")
	self._nodeAvatar = ccui.Helper:seekNodeByName(self._target, "NodeAvatar")
    self._imageAvatar:setVisible(false)
end

function CommonStoryAvatar:bind(target)
	self._target = target
    self:_init()
    cc.setmethods(target, self, EXPORTED_METHODS)
end

function CommonStoryAvatar:unbind(target)
    cc.unsetmethods(target, EXPORTED_METHODS)
end

function CommonStoryAvatar:updateUI(heroId, limitLevel, limitRedLevel)
    local param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_HERO, heroId, nil, nil, limitLevel, limitRedLevel)
    local resData = param.res_cfg
    if resData.story_res_spine ~= 0 then
        self:_createHeroSpine(resData.story_res_spine)
        self._imageAvatar:setVisible(false)
    else
        self:_createHeroImage(resData.story_res)
        if self._spine then
            self._spine:setVisible(false)
        end
    end
end

function CommonStoryAvatar:_createHeroSpine(spineId)
    if not self._spine then
        self._spine = require("yoka.node.SpineNode").new(1, cc.size(1024, 1024))
        self._nodeAvatar:addChild(self._spine)
    end
    self._spine:setAsset(Path.getStorySpine(spineId))
    self._spine:setAnimation("idle", true)
    self._spine:setVisible(true)
end

function CommonStoryAvatar:_createHistorySpine(spineId)
    if not self._spine then
        self._spine = require("yoka.node.SpineNode").new(1, cc.size(1024, 1024))
        self._nodeAvatar:addChild(self._spine)
    end
    self._spine:setAsset(Path.getSpine(spineId))
    self._spine:setAnimation("idle", true)
    self._spine:setVisible(true)
end

function CommonStoryAvatar:_createHeroImage(imageId)
    local imgPath = Path.getChatRoleRes(imageId)
    self._imageAvatar:ignoreContentAdaptWithSize(true)
    self._imageAvatar:loadTexture(imgPath)
    self._imageAvatar:setVisible(true)
end

function CommonStoryAvatar:updateUIByResId(resId)
    local resData = HeroRes.get(resId)
    if resData.story_res_spine ~= 0 then
        self:_createHistorySpine(resData.story_res_spine)
        self._imageAvatar:setVisible(false)
    else
        self:_createHeroImage(resData.story_res)
        if self._spine then
            self._spine:setVisible(false)
        end
    end   
end

return CommonStoryAvatar